### DISCUSSION TITLE
Outlook roaming signatures

### DISCUSSION THREAD

**Discussion by jpvj** [Set-OutlookSignatures/Set-OutlookSignatures#62] [Ideas]
Hello,

I stumbled across your script and did some testing and so far it looks good :-)

As many other users we are hit by the roaming signature update problem.
I found discussion - https://petri.com/outlook-for-windows-cloud-signatures/ :

"akoepke
25 November, 2022

I found they still work if you can change the file names to end with the account name. So ours have become "username (email).htm" instead of just "username.htm".

This is then associated with the correct email account."

As far as I can see it seems to import the signature in Outlook after renaming the file in %appdata%\Microsoft\Signatures\ from signature.htm to "signature (email).htm".

**16.25. What about the roaming signatures feature announced by Microsoft?**

I haven't done extensive testing so I don't know if it only fixes the new way of Outlook handling signatures or Outlook will take care of the roaming part too, but it sure looks as I can attach the signatures to the mailbox as default by this simple name change.

**Comment by GruberMarkus** [comment]
Hi @jpvj,

this Outlook behavior would be a significant change compared to beta versions from a few weeks ago. I will test this in the following days.

While it is already clear through reverse engineering how Outlook communicates with the cloud regarding signatures, there are open questions how Outlook actually handles the data internally.
Also, Outlook currently does not communicate with the cloud via the Graph but the Substrate API, and Microsoft has not released any hints how to correctly integrate the feature into own applications.

It seems already quite clear that Set-OutlookSignatures will only support cloud roaming signatures for the personal mailbox of the logged in user. This is because it is not yet clear if shared mailboxes will really support roaming signatures, and because storing personalized signatures in shared mailboxes does not make much sense anyhow.

In contrast, I think about an option storing all signatures, including signatures for shared mailboxes, in the mailbox of the currently logged in user. This would be handy when a user accesses shared mailboxes in Outlook Web.
Would you find such an option useful?

**Reply by jpvj** [reply]
Hi Markus,

Thank you for your quick reply and sharing your knowledge about the (lack of) API.

As far as I understand most - if not all - products on the markets relies on DisableRoamingSignaturesTemporaryToggle = 1. This totally supports your statement about the missing API.

I understand you approach to be "Set-OutlookSignature will store the signatures in the mailbox (when the API will be made available)". Correct?

Regarding the shared mailboxes - that's a really good question. There are probably requests for both "general" (e.g. departmentwise) or personal signatures. Whether the "general" signature stored in the shared mailbox is possible time will tell. So far I only did a test with an automapped mailbox and you can definitively not set a signature for an automapped shared mailbox i O365. General signatures should follow the shared mailbox if ever possible. If using your personal signature in the shared mailbox works, that would be fine for me.

**Reply by jpvj** [reply]
PS: If it's clear how Outlook communicates with the could, would it be possible to reimplement this in the script? 
I would expect Outlook to read the signatures from the mailbox - just as it was written by another Outlook application.

Do you by chance have any links to where the reverse engineered communication is documented?

Thank you in advance!

**Comment by GruberMarkus** [comment]
Unfortunately, the tests with the most recent beta version of Outlook (Version 2212 Build 16.0.15917.20000) are negative.

When creating a signature in the file system and naming it "test (user@example.com).htm", everything looks fine in Outlook. Unfortunately, changes are not synchronized with the mailbox itself. Restarting Outlook or letting Outlook run for at least one hour does not help.

The signature synchronization is only triggered when
- Signatures are changed in Outlook Web and Outlook triggers a synchronization (restart of Outlook are presumably some sort of random timer)
- Signatures are changed via the Outlook GUI.

In either case, the changes in the file system are reverted or deleted, only the cloud settings count.

Either I find a way to alter the registry values in 'HKCU\Software\Microsoft\Office\Outlook\Settings\Data' starting with 'user@example.com', or I wait until an official API is available. I currently prefer the latter.

**Reply by jpvj** [reply]
Thank you for testing.

I agree - the API would definitively be preferrable.

Hacking the registry and triggering Outlook to perform the sync could be an option.
Did you try to monitor the registry, so see what changes Outlook does? Maybe you could catch a flag indicating "ready to sync" / "synchronized"?

**Comment by GruberMarkus** [comment]
I have done the reverse engineering myself by watching https traffic of Outlook and Outlook Web, and by monitoring the registry. The picture is quite good already in terms of how to communicate with the cloud, but the local transforms that are obviously made are still a black box.

Also, the API calls have changed several times in the last 12 months.

Until there is no more doubt how this really works, I will not implement cloud signatures in Set-OutlookSignatures - the risk to damage something is just too high.

In the meantime, you have to rely on DisableRoamingSignaturesTemporaryToggle=1 and have signature roaming disabled at the tenant level via Microsoft support.

**Reply by jpvj** [reply]
Thank you for explaining.

I more or less came to the same conclusion this morning - we will rely on the supported registry change.
Would you mind explaining why it have to be disabled on the tenant level too?
(It's a bit more complicated here, as we are only 1.000 users out of 30.000 sharing the tenant)

**Comment by GruberMarkus** [comment]
Set-OutlookSignatures can deploy one (and only one) signature to Outlook Web. This feature does not work, when the new roaming signature feature is enabled at the tenant level (this has nothing to do with the DisableRoamingSignaturesTemporaryToggle registry value on a client).

As far as I know, the new feature can not be enabled or disabled at the mailbox level, only at the tenant level. And this configuration can only be done by Microsoft, not by a tenant admin.

In README, the FAQ 'What about the roaming signatures feature announced by Microsoft?' describes the problem in general. The sub chapter "Please be aware of the following problem" focuses on the problem setting the Outlook Web signature via Set-OutlookSignatures. It describes the problem and two possible workarounds.
Another choice would be to no longer have Set-OutlookSignatures set a signature in Outlook Web by using the 'SetCurrentUserOutlookWebSignature false' parameter.

**Comment by RaphRi** [comment]
Hello, in a need to centrally manage signatures in our school with O365 tenant, I found your  project wich is very good !
However, my first tests with the cmdlet Set-MailboxMessageConfiguration doesn't seems to work anymore beacause of Roaming profile activated.
At least CodeTwo seems to have manage to use the new apis  : https://www.youtube.com/watch?v=Dd4SOA5dUmw&feature=youtu.be&t=1352 and https://office365itpros.com/2020/09/25/outlook-signature-api/
But I prefer opensource :) ! 
Do you know if there is any availibility/documententation on the new API or is it still obscured by Microsoft ?

Else I'm thinking of asking Microsoft support to disable the roaming feature...

Thanks !

**Reply by GruberMarkus** [reply]
Hi @RaphRi,

I am in direct contact with the Microsoft product group. There is no public API available. CodeTwo does not have one, too (their modern Outlook Web Add-In uses a completely different technology based on API calls to their own signature cloud infrastructure).

Even Microsoft internally, the API is still undergoing changes - for example, the most recent versions of Office Click to run suddenly can't handle roaming signatures anymore... And it's still not integrated in Graph, but "only" in the underlying Office Substrate API, which is at least one layer below the Graph API and usually not available for public access at all.

It is a long and painfuly story. Although i reverse engineered three versions of the roaming feature - or rather: because I did - I will wait until Microsoft publishes an offical API:

Until then, you may want to take a few minutes to read through my thoughts on this in the README: https://github.com/GruberMarkus/Set-OutlookSignatures#1625-what-about-the-roaming-signatures-feature-in-exchange-online

**Reply by jpvj** [reply]
Thank you very much for the update. 
It so weird to observe how bad this is handled by MS.  Seems like a non-finished feature was released my MS. 

It’s going to be interesting how it ends 😳

**Reply by RaphRi** [reply]
Thank you very much for the detailed status. I'll open a support to get back to previous release and wait public availability of some API. If there will be some one day...

**Reply by jpvj** [reply]
Hi @GruberMarkus,

Just wondering if you have heard anything new from MS?

We would love to have the script working, so just consider this as "the little boy waiting for his Christmas present" :-)

**Reply by GruberMarkus** [reply]
Unfortunately, there is nothing new to report.

There is still:
- No public API (even the lates Outlook version do not use Graph)
- No end date for the rollout to cloud tenants (according to the Microsoft 365 Roadmap, the rollout is still at 50 %, where it is stuck for several months)
- No definition how roaming signatures are handled for shared mailboxes

Even worse, latest Outlook releases have problem with roaming signatures: Showing each signature three times in the signature editor window, only downloading but not uploading signatures for certain mailboxes, uploading signatures to mailboxes where roaming is not yet enabled.

The best thing to do at the moment is to either wait for the Christmas present, or - what I usually recommend -: Disable roaming signatures on the client side and continue your rollout - roaming signatures are then just an additional goodie.

**Comment by GruberMarkus** [comment]
Update: Set-OutlookSignatures supports roaming signatures since v4.0.0. Enjoy!

**Reply by jpvj** [reply]
Great news!!!

Would you care to "blog" a little about how/what/when? :-)

**Reply by GruberMarkus** [reply]
I can not go into any details, as the public API has not been released by Microsoft.

Please look for "MirrorLocalSignaturesToCloud" in the README file of the current version to read about the details I am allowed to share.

**Reply by GruberMarkus** [reply]
The demo video mentions the feature briefly and shows the results: https://www.youtube.com/watch?v=K9TrCjTdRUI&t=815s

**Reply by RaphRi** [reply]
Thanks very much for that. The project was on hold as I attempted to disable roaming signatures with Microsoft support which never succeeded. Ticket closed with the redirect to one on premise Outlook app solution. Seriously ?

**Reply by GruberMarkus** [reply]
You no longer need Microsoft to disable roaming signatures, because you have two alternatives:

Use Set-OutlookSignatures with the MirrorLocalsignaturesToCloud parameter, or disable roaming signatures yourself: The Exchange cmdlet Set-OrganizationConfig now has the PostponeRoamingSignaturesUntilLater parameter (details on https://learn.microsoft.com/en-us/powershell/module/exchange/set-organizationconfig?view=exchange-ps).

**Reply by GruberMarkus** [reply]
As a personal note: I think that roaming signatures are the future, and I do not want to miss them, even when currently only Outlook on Windows, Outlook Web and New Outlook support them. I am convinved that as soon as the public API is published, more and more email clients (especially mobile apps) will support this feature.

**Reply by RaphRi** [reply]
Indeed since It's working with roaming signatures, no need to disable it. Thanks for the great work !

**Comment by GruberMarkus** [comment]
My pleasure!

**Comment by 2PintChristianN** [comment]
[Set-MailboxMessageConfiguration](https://learn.microsoft.com/en-us/powershell/module/exchange/set-mailboxmessageconfiguration) Now has `-SignatureHtmlBody` and `-SignatureName`

[Get-MailboxMessageConfiguration](https://learn.microsoft.com/en-us/powershell/module/exchange/get-mailboxmessageconfiguration) also has `-SignatureName` and returns some interesting props: `SignaturesList`, `SignatureName`, `SignatureHtmlBody`
Suggesting that this is now official, but documentation is still lacking. I have not tested these yet, as we are still running with roaming disabled.
